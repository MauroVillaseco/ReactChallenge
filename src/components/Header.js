import React from "react";
import styled from 'styled-components';


const AppHeader = styled.header`
 background-color: #DDD;
 font-size: 28px;
 padding: 20px;
 color: #000;
 border-bottom: 1px solid #000;
`;


export const Header = () => (

    <AppHeader>Challenge By Mauro</AppHeader>
);
