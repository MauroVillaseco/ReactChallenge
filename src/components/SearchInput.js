import React from 'react';
import styled from 'styled-components'
import PropTypes from "prop-types";
const BarStyling = styled.input`
    height: 25px;
    background: inherit;
    border: none;
    border-bottom: 1px solid;
    border-radius: 0;
      &:focus {
        outline: none;
      }
  `;
  const CrossStyles = styled.span`
    cursor: pointer;
    left: -12px;
    position: relative;
    opacity:0;
    pointer-events:none;
    transition: opacity .3s ease-in-out;
    &.active {
      opacity:1;
      pointer-events:auto;
    }
  `;
export const SearchInput = ({input:keyword, onChange:setKeyword}) => {
  return (
    <div>
      <BarStyling 
      key="random1"
      value={keyword || ''}
      placeholder={"Search People.."}
      onChange={(e) => setKeyword(e.target.value)}
      />
      <CrossStyles className={keyword && 'active'} onClick={() => setKeyword('')}>x</CrossStyles>
    </div>
  );
}


Element.propTypes = {
  keyword: PropTypes.string.isRequired
};

