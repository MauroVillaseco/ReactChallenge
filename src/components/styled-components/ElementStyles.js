import myImage from '../../img/editIcon.svg';
import styled from 'styled-components'
import * as palette from '../../styles/styledComponentsGenerals.js';

export const ParentElement = styled.div`
background-color: ${palette.WHITE};
color: ${palette.BLACK};
border-radius:3px;
display: grid;
text-align: center;
box-shadow: 0px 4px 9px -1px rgba(61,61,61,1);
background: linear-gradient(to bottom, #416dff 30%, ${palette.WHITE} 30%);
height: 370px;
grid-template-columns: 75%;
justify-content: center;
position: relative;
`;

export const TextContainer = styled.div`
padding-bottom: 20px;
color: ${palette.BLACK};
border-radius:3px; 
display: inline-block;
align-self: end;
`;

export const TextContainerDouble = styled.div`
padding-bottom: 20px;
color: ${palette.BLACK};
border-radius:3px;
align-self: end;
display: flex;
justify-content: center;
padding-bottom: 8px !important;
input {
max-width: 80px;
    &:first-of-type{
      margin-right: 25px;
    }
  }
`;

export const HeadElement = styled.h4`
font-size: 20px;
color: ${palette.WHITE};
text-transform: capitalize;
text-overflow: ellipsis;
overflow: hidden;
`;

export const TextElement = styled.p`
font-size: 15px;
color: #878787;
text-overflow: ellipsis;
overflow: hidden;
`;

export const ImageContainer = styled.div`
  width: 130px;
  height: 130px;
  position: relative;
  overflow: hidden;
  border-radius: 50%;
  border: 3px ${palette.WHITE} solid;
  box-shadow: 0px 4px 9px -1px rgba(61,61,61,1);
  position: relative;
  left: 50%;
  transform: translateX(-50%);
  img {
      display: inline;
      margin: 0 auto;
    }
`;

export const IconElement = styled.div`
position: absolute;
left: 15px;
top: 15px;
background-image: url(${myImage});
filter: invert(100%);
-webkit-filter: invert(100%);
width: 20px;
height: 20px;
background-size: 20px 20px;
cursor:pointer;
transition: opacity .3s ease-in-out;
&:hover {
 opacity: 0.8;
}
`;
export const InputEditable = styled.input`
height: 25px;
background: inherit;
border: none;
border-bottom: 1px solid;
border-radius: 0;
 &:focus {
   outline: none;
 }
`;

export const ButtonForm = styled.button`
 background: #416dff;
 height: 30px;
 cursor: pointer;
 border: 1px solid #ccc;
 border-radius: 3px;
 color:${palette.WHITE};
 font-weight: bold;
`;
