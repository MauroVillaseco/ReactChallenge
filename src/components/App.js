import React from "react";
import { Header } from "./Header";
import { VisibleElementList } from "../containers/VisibleElementList";
import { SortElements } from "../containers/SortElements";
import '../styles/main.scss';


const App = () => {
  return (
  <div>
    <Header />
    <SortElements />
    <VisibleElementList />
  </div>
)};

export default App;