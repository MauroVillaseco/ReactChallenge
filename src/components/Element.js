import React, { useState } from "react";
import PropTypes from "prop-types";
import { useElementsStore } from "../store";
import {ParentElement,IconElement, TextContainer, TextContainerDouble,InputEditable, ImageContainer,ButtonForm,HeadElement,TextElement } from "./styled-components/ElementStyles"


export const Element = ({ first, last, email, phone, picture, location }) =>{
  const [ editable, setEditable ] = useState();
  const { editElement } = useElementsStore();

  const onSubmit = (evt) => {
    evt.preventDefault();
    console.log(evt)
    editElement({'first':evt.target.first.value,'last':evt.target.last.value, 'email': evt.target.email.value,'phone': evt.target.phone.value, 'city': evt.target.city.value, 'state':evt.target.state.value}, evt.target.email.defaultValue);
    setEditable(false)
  }

if (editable) {
  return (
    <form className='form-submit' onSubmit={onSubmit}>
      <ParentElement >
        <IconElement data-testid="cancel-edit" onClick={() => setEditable(false)}></IconElement>
        <TextContainerDouble className='double-input'>
        <InputEditable className='color-white' name='first' type='text' defaultValue={first} maxLength='30' required/> 
        <InputEditable className='color-white' name='last' type='text' defaultValue={last} maxLength='30' required/>
        </TextContainerDouble>    
        <ImageContainer className='image-cropper'>    
          <img alt='Predefined' src={picture.large}/>
        </ImageContainer>
        <TextContainer>
          <InputEditable type='email' name='email' maxLength='50' placeholder='Enter Email..' defaultValue={email} required/>
        </TextContainer>
        <TextContainer>
          <InputEditable name='phone' maxLength='50' defaultValue={phone} placeholder='Enter Phone..' required/>
        </TextContainer>
        <TextContainerDouble className='double-input'>    
          <InputEditable name='city' type='text' maxLength='80' defaultValue={location.city} placeholder='Enter City..' required/>
          <InputEditable name='state' type='text' maxLength='80' defaultValue={location.state} placeholder='Enter State..' required/>
        </TextContainerDouble>
        <ButtonForm data-testid="submit-edit" type='submit'>Save Changes</ButtonForm>
      </ParentElement>
    </form>
  );
}
return (
 
  <ParentElement>
    <IconElement data-testid="edit" onClick={() => setEditable(true)}></IconElement>
    <TextContainer>
    <HeadElement>{first} {last}</HeadElement>
    </TextContainer>    
    <ImageContainer className='image-cropper'>    
      <img alt='Predefined' src={picture.large}/>
    </ImageContainer>
    <TextContainer>
      <TextElement>{email}</TextElement>
    </TextContainer>
    <TextContainer>
      <TextElement>{phone}</TextElement>
    </TextContainer>
    <TextContainer>    
      <TextElement>{location.city}, {location.state}</TextElement>
    </TextContainer>
  </ParentElement>
);

}
Element.propTypes = {
  first: PropTypes.string.isRequired,
  last: PropTypes.string.isRequired,
  picture: PropTypes.object.isRequired,
  email: PropTypes.string.isRequired,
  phone: PropTypes.string.isRequired,
  location: PropTypes.object.isRequired
};
