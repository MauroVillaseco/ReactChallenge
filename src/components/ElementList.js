import React from "react";
import PropTypes from "prop-types";
import { Element } from "./Element";
import styled from 'styled-components'
const ContainerElement = styled.div`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
  grid-gap: 20px;
  align-items: stretch;
  margin-top: 20px;
`;

// List that recives all the elements and iterate between them to pass the data to the element Component.
export const ElementList = ({ elements }) => {
  if (!elements) { 
    return null;
  }
  return(
 
  <ContainerElement>
       
      {elements.map(element => (
        //i couldn't use the user id here, because sometimes it was null
        <Element key={element.registered.date}
        first={element.name.first}
        last={element.name.last}
        email={element.email}
        phone={element.phone}
        picture={element.picture}
        location={element.location}
          />
      ))}
    
  </ContainerElement>   
  )
};

ElementList.propTypes = {
  elements: PropTypes.array.isRequired
};
