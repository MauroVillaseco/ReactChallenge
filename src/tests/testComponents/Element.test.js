import React from 'react'
import { Element}  from '../../components/Element'
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import '@testing-library/jest-dom/extend-expect' 
import Enzyme ,{ mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { IconElement } from "../../components/styled-components/ElementStyles" ;
import { TestExampleDataStore } from "../_helpers/testHelpers" ;
Enzyme.configure({ adapter: new Adapter() });
const mockStore = configureStore([]);
describe('actions', () => {
let store;
let component;

beforeEach(() => {
    store = mockStore(TestExampleDataStore)
    const element = { first :'Mauro', last:'last', email:'email@email.com', phone:'123456', picture: {large: 'https://randomuser.me/api/portraits/men/83.jpg'},
    location:{ city: 'Mar del plata', state: 'Buenos Aires'} }

    store.dispatch = jest.fn();

    // I try to use react-test-renderer for simple testing 
    component = renderer.create(
        <Provider store={store}>
        <Element first={element.first}
        last={element.last}
        email={element.email}
        phone={element.phone}
        picture={element.picture}
        location={element.location} />
        </Provider>
    );
});
    it('should render with given state from Redux store', () => {
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should show the inputs to edit on edit button click', () => {
        renderer.act(() => {
            component.root.findByProps({'data-testid': "edit"}).props.onClick();
        });        
        expect(component.root.findByProps({'data-testid': "cancel-edit"}).props).toBeTruthy();
        
    });
    it('should show the cards', () => {
        expect(component.root.findByProps({'data-testid': "edit"}).props).toBeTruthy();
    });
    it('should test button submit is present', () => {
        const element = { first :'Mauro', last:'last', email:'email@email.com', phone:'123456', picture: {large: 'https://randomuser.me/api/portraits/men/83.jpg'},
        location:{ city: 'Mar del plata', state: 'Buenos Aires'} };
        // I use Enzyme here 
        const elementComponent = mount(
            <Provider store={store}>
                <Element 
                first={element.first}
                last={element.last}
                email={element.email}
                phone={element.phone}
                picture={element.picture}
                location={element.location}/>
            </Provider>);
        elementComponent.find(IconElement).simulate('click');
        expect(elementComponent.find('.form-submit').length).toBe(1);
        expect(store).toBeTruthy();
        expect(elementComponent).toMatchSnapshot();
    });
})