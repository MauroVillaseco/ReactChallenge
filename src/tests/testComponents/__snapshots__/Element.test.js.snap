// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should render with given state from Redux store 1`] = `
<div
  className="sc-bdfBwQ gmQNOd"
>
  <div
    className="sc-jSgupP bukQAx"
    data-testid="edit"
    onClick={[Function]}
  />
  <div
    className="sc-gsTCUz gaZGJg"
  >
    <h4
      className="sc-dlfnbm fSmEn"
    >
      Mauro
       
      last
    </h4>
  </div>
  <div
    className="sc-eCssSg bJgeTb image-cropper"
  >
    <img
      alt="Predefined"
      src="https://randomuser.me/api/portraits/men/83.jpg"
    />
  </div>
  <div
    className="sc-gsTCUz gaZGJg"
  >
    <p
      className="sc-hKgILt fRAooq"
    >
      email@email.com
    </p>
  </div>
  <div
    className="sc-gsTCUz gaZGJg"
  >
    <p
      className="sc-hKgILt fRAooq"
    >
      123456
    </p>
  </div>
  <div
    className="sc-gsTCUz gaZGJg"
  >
    <p
      className="sc-hKgILt fRAooq"
    >
      Mar del plata
      , 
      Buenos Aires
    </p>
  </div>
</div>
`;

exports[`actions should test button submit is present 1`] = `ReactWrapper {}`;
